server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java

```diff
@@ -0,0 +1,273 @@
+                if (limit <= 0 || limit > MAX_LIMIT) {
+                    throw new IllegalArgumentException("Invalid value for parameter k for top-k aggregation");
+                }
+                int maxMapSize = maxMapSize(limit);
+                ramAccounting.addBytes(calculateRamUsage(maxMapSize));
+                DataType<?> dataType = DataTypes.guessType(value);
+                state = new TopKState(new ItemsSketch<>(maxMapSize), dataType, limit);
+            } else if (args.length == 1) {
+                // We don't have a limit provided, let's go with the default values
+                DataType<?> dataType = DataTypes.guessType(value);
+                ramAccounting.addBytes(calculateRamUsage(DEFAULT_MAP_SIZE));
+                state = new TopKState(new ItemsSketch<>(DEFAULT_MAP_SIZE), dataType, DEFAULT_LIMIT);
+            }
+        }
+        if (state instanceof TopKState topKState) {
+            topKState.itemsSketch.update(value);
+        }
+        return state;
+    }
+
+    private static int maxMapSize(int x) {
+        // max map size should be 4 * the limit based on the power of 2 to avoid errors
+        return (int) Math.pow(2, Math.ceil(Math.log(x) / Math.log(2))) * 4;
+    }
+
+    private static long calculateRamUsage(long maxMapSize) {
+        // The internal memory space usage of item sketch will never exceed 18 * maxMapSize bytes, plus a small
+        // constant number of additional bytes.
+        // https://datasketches.apache.org/docs/Frequency/FrequentItemsOverview.html
+        return maxMapSize * 18L;
+    }
+
+    @Override
+    public State reduce(RamAccounting ramAccounting, State state1, State state2) {
+        if (state1 instanceof TopKState t1 && state2 instanceof TopKState t2) {
+            return new TopKState(t1.itemsSketch.merge(t2.itemsSketch), t1.dataType, t1.limit);
+        } else if (state1 instanceof Empty) {
+            return state2;
+        } else {
+            return state1;
+        }
+    }
+
+    @Override
+    public List<Object> terminatePartial(RamAccounting ramAccounting, State state) {
+        if (state instanceof TopKState topKState) {
+            ItemsSketch.Row<Object>[] frequentItems = topKState.itemsSketch.getFrequentItems(ErrorType.NO_FALSE_NEGATIVES);
+            int limit = Math.min(frequentItems.length, topKState.limit);
+            var result = new ArrayList<>(limit);
+            for (int i = 0; i < limit; i++) {
+                var item = frequentItems[i];
+                result.add(Map.of("item", item.getItem(), "frequency", item.getEstimate()));
+            }
+            return result;
+        }
+        return List.of();
+    }
+
+    public DataType<?> partialType() {
+        return StateType.INSTANCE;
+    }
+
+    sealed interface State {
+        State EMPTY = new Empty();
+    }
+
+    record Empty() implements State { }
+
+    record TopKState(ItemsSketch<Object> itemsSketch, DataType<?> dataType, int limit) implements State { }
```

Comment:
I think moving the dataType into the state shouldn't be necessary as the aggregation itself contains it as part of the boundSignature.
Would make the payload smaller

Reply:
I need the datatype for serialization because the bound signature is unavailable in DataType/Streamer for State.

Reply:
maxMapSize and limit are not the same. The limit is what the user wants to get back as top-k. maxMapSize defines the window of the top-k tracking. maxMapSize is usually factor 3-5 of the limit in most implementations because you want to track more items than you return.

Reply:
If maxMapSize is too small the error increases a lot, because the tracking gets inaccurate.